telepresence/proxy/operation.py

```diff
@@ -0,0 +1,312 @@
+        ("expose", PortMapping),
+        ("env", Dict[str, str]),
+        ("service_account", str),
+    ]
+)
+
+
+class ProxyOperation:
+    def __init__(self, intent: ProxyIntent) -> None:
+        self.intent = intent
+
+    def prepare(self, runner: Runner) -> None:
+        pass
+
+    def act(self, _: Runner) -> RemoteInfo:
+        raise NotImplementedError()
+
+
+LegacyOperation = Callable[[Runner, str, PortMapping, Dict[str, str], str],
+                           Tuple[str, Optional[str]]]
+
+
+class Legacy(ProxyOperation):
+    def __init__(self, intent: ProxyIntent, legacy_op: LegacyOperation):
+        super().__init__(intent)
+        self.op = legacy_op
+
+    def act(self, runner: Runner) -> RemoteInfo:
+        deployment_arg = self.intent.name
+        if self.intent.container:
+            deployment_arg += ":" + self.intent.container
+
+        tel_deployment, run_id = self.op(
+            runner,
+            deployment_arg,
+            self.intent.expose,
+            self.intent.env,
+            self.intent.service_account,
+        )
+
+        remote_info = get_remote_info(
+            runner,
+            tel_deployment,
+            "unused, right?",
+            run_id=run_id,
+        )
+
+        return remote_info
+
+
+def create_with_cleanup(runner: Runner, manifests: Iterable[Manifest]) -> None:
+    kinds = set(str(manifest["kind"]).capitalize() for manifest in manifests)
+    kinds_display = ", ".join(kinds)
+    manifest_list = make_k8s_list(manifests)
+    manifest_json = json.dumps(manifest_list)
+    try:
+        runner.check_call(
+            runner.kubectl("create", "-f", "-"),
+            input=manifest_json.encode("utf-8")
+        )
+    except CalledProcessError as exc:
+        raise runner.fail(
+            "Failed to create {}:\n{}".format(kinds_display, exc.stderr)
+        )
+
+    def clean_up() -> None:
+        runner.show("Cleaning up {}".format(kinds_display))
+        runner.check_call(
+            runner.kubectl(
+                "delete",
+                "--ignore-not-found",
+                "--wait=false",
+                "--selector=telepresence=" + runner.session_id,
+                ",".join(kinds),
+            )
+        )
+
+    runner.add_cleanup("Delete proxy {}".format(kinds_display), clean_up)
+
+
+class New(ProxyOperation):
+    def prepare(self, runner: Runner) -> None:
+        self.manifests = []  # type: List[Manifest]
+
+        # Construct a Pod manifest
+        pod = make_new_proxy_pod_manifest(
+            self.intent.name,
+            runner.session_id,
+            get_image_name(runner, self.intent.expose),
+            self.intent.service_account,
+            self.intent.env,
+        )
+        self.manifests.append(pod)
+
+        # Construct a Service manifest as needed
+        if self.intent.expose.remote():
+            svc = make_svc_manifest(
+                self.intent.name,
+                dict(telepresence=runner.session_id),
```

Comment:
There doesn't appear to be a need to vary labels from selectors, maybe accept a map of labels and reuse it?